pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -135,36 +148,55 @@ func (sf *IteratorDatastore) ReadStartingWithUser(
 
 	if found {
 		keyItem.counter++
+		keyItem.iter.mu.Lock()
+		defer keyItem.iter.mu.Unlock()
 		sf.internalStorage.mu.Unlock()
 		span.SetAttributes(attribute.Bool("found", true))
 
-		// For now, this will be identical to the not found path.  When the SF iterator datastore
-		// is fully implemented, the found path will look up the shared iterator and create
-		// a clone from it.
-	} else {
-		if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
-			sf.internalStorage.mu.Unlock()
-			sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadStartingWithUser).Inc()
-			// we cannot share this iterator because we have reached the size limit.
-			return sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
-		}
+		sharedIteratorQueryHistogram.WithLabelValues(
+			storagewrappersutil.OperationReadStartingWithUser, strconv.FormatBool(found),
+		).Observe(float64(time.Since(start).Milliseconds()))
 
-		sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
-			counter: 1,
+		item, err := keyItem.iter.clone()
+		if err != nil {
+			go func() {
+				_, _ = sf.deref(cacheKey)
+			}()
+			return nil, err
 		}
+		return item, nil
+	}
+	if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
+		span.SetAttributes(attribute.Bool("overlimit", true))
+
 		sf.internalStorage.mu.Unlock()
-		span.SetAttributes(attribute.Bool("found", false))
+		sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadStartingWithUser).Inc()
+		// we cannot share this iterator because we have reached the size limit.
+		return sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+	}
+
+	newIterator := newSharedIterator(sf, cacheKey, nil, sf.maxAliveTime)
+	newIterator.mu.Lock()
+	defer newIterator.mu.Unlock()
+	sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
+		counter: 1,
+		iter:    newIterator,
 	}
+	sf.internalStorage.mu.Unlock()
+	span.SetAttributes(attribute.Bool("found", false))
+
 	actual, err := sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
 	if err != nil {
+		newIterator.queryErr = err
 		_, _ = sf.deref(cacheKey)
 		return nil, err
 	}
+	newIterator.inner = actual
```

Comment:
So if 2 queries for the same iterator come at the same time:
1. Whichever grabs the lock first will generate the entry into `internalStorage`.
2. It will yield the lock allowing the second to go into `if found` and yield that.
3. Then the first one which grabbed the lock will attempt to build _actual_, but if it fails, there will be no `inner`.

If I'm not wrong, it will blow up in that case or am I missing something there?

Reply:
I will need to move sf.internalStorage.mu.Unlock() to be after the query

Reply:
I see what you are implying - in this case, I will update it so that it is not using the shared iterator at all.